### ISSUE TITLE
Redis DB is full of unprocessed events

### ISSUE THREAD

**Issue by sander85** [getsentry/self-hosted#1787]
### Self-Hosted Version

22.10.0

### CPU Architecture

x86_64

### Docker Version

20.10.12

### Docker Compose Version

1.29.2

### Steps to Reproduce

1. Just use Sentry

### Expected Result

Redis shouldn't have too many keys in the database.

### Actual Result

Since upgrading to 22.10.0 we see a lot of keys in the redis database. And the number grows pretty fast if DB is not flushed. This didn't happen with previous version of Sentry. Many of the keys started with c: or something like that. I opened one and it seemed like an issue with event_id that I couldn't find from the Sentry UI. So it seems that some component is not picking up the keys from redis to transfer them to permanent storage or something like that. Our PostgreSQL and Redis are configured outside of the docker-compose setup. But some of the new events are saved and can be seen in the UI. So not all of the events go missing.

**Comment by sander85** [comment]
Forgot to add that I can't see anything interesting from any of the logs that I can access. No errors nor warnings. Just randomly part of the events remain in redis.

**Comment by hubertdeng123** [comment]
Thanks for bringing this up, will have to do some investigation here

**Comment by sander85** [comment]
The issue hasn't been fixed. I don't have any more info to input either as I don't know which part is responsible for cleaning up redis.

**Comment by igorcoding** [comment]
We started to face the same issue on version 22.8.0 for no reason whatsoever. No upgrades, no downtime, just for some reason redis started consuming more and more RAM. I confirm that the problem is that event keys are not being deleted. Unfortunately, I cannot understand what had caused this issue, because there are no errors or warning in logs in sentry processes. It looks like some sort of failure in tasks processing, but again - can't get to the root of it.
Here is output of `redis-cli --bigkeys`:
![telegram-cloud-photo-size-2-5188550949084447672-y](https://user-images.githubusercontent.com/5521116/203900968-4123d36c-8844-4b2b-8c4b-d0eca99ecf1f.jpg)
I've been thinking that maybe try to update sentry, but as I can see the problem is with newer versions as well.

**Comment by TheBlusky** [comment]
Same issue here. Is there any workaround / task to run tu free memory ?

**Comment by robkorv** [comment]
> Same issue here. Is there any workaround / task to run to free memory ?

See https://github.com/getsentry/self-hosted/issues/1796#issuecomment-1304859942

**Comment by le0pard** [comment]
we start getting same issue with redis after update from 22.12.0 to 23.2.0 (on this update we switched to aws s3 for Node Storage)

**Comment by hheexx** [comment]
Same problem here happening again on 23.2.0 also

**Comment by namco1992** [comment]
The fix here https://github.com/getsentry/self-hosted/pull/1817 fixed our problem, maybe you can try and see if it works for you.

**Comment by le0pard** [comment]
@namco1992 it was merged, so it already in yml file, which was fetched for 23.2.0 update

**Comment by hheexx** [comment]
This one worked for me:
https://github.com/getsentry/self-hosted/issues/1796#issuecomment-1304859942

**Comment by le0pard** [comment]
Yes, but this is not a fix for bug. It just hide problem by removing redis keys, when it is not time to remove it (but we don't have space for new keys)

**Comment by le0pard** [comment]
I did some monitoring for redis. Hard to say the reason of memory growth (too many events), but here interesting pattern:

Insert some keys with pattern "c:1:e:<hash>:<number>"
```
"SETEX" "c:1:e:aa4223d9fb9b4acf953868aa04d00e87:46" "86400" "{\"event_id\":\"aa4223d9fb9b4acf953868aa04d00e87\"
```

After this get and delete data by key with additional prefix ":a"

```
"GET" "c:1:e:aa4223d9fb9b4acf953868aa04d00e87:46:a"
"DEL" "c:1:e:aa4223d9fb9b4acf953868aa04d00e87:46:a"
```

After some time some process read this key:

```
"GET" "c:1:e:aa4223d9fb9b4acf953868aa04d00e87:46"
```

But nobody cleanup it. So we need wait 24 hours based on "setex" value to redis itself cleanup this events. And this events is not small, sometimes ~200 kbytes value

Here number of such keys growing over time in redis

```
$ docker-compose exec redis redis-cli --scan --pattern 'c:1:e:*' | wc -l
162070
$ docker-compose exec redis redis-cli --scan --pattern 'c:1:e:*' | wc -l
162136
$ docker-compose exec redis redis-cli --scan --pattern 'c:1:e:*' | wc -l
162205
```

**Comment by darklow** [comment]
If you don't really care about anything unprocessed and encounter this issue when redis eats up all memory and eventually crashes whole on-premise deployment, what helped me was following:
```
docker exec -it sentry-self-hosted_redis_1 redis-cli flushall
```
which basically deletes anything on redis. we had couple of projects creating hundreds of k of issues which lead to disk space issues and after resize sentry kept crashing until I noticed that it is redis that causes memory to be filled up and cleaning it all helped.

**Comment by azaslavsky** [comment]
> docker exec -it sentry-self-hosted_redis_1 redis-cli flushall doesn't work for me.

Hmm, just tried this on a local instance and it seemed to work? Have you tried scrubbing the docker volume backing redis?

**Comment by meotimdihia** [comment]
@azaslavsky I could be mistaken it looks like the command `docker exec -it sentry-self-hosted_redis_1 redis-cli flushall` works again for me.

**Comment by longwave** [comment]
We hit the same symptom on self-hosted 26.8.0, but the cause was a task backlog, not a leak.

To tell which one you have, check the task queue lag:

```
docker compose exec kafka kafka-consumer-groups --bootstrap-server localhost:9092 --describe --group taskworker
```

Ours was about 2.1 million. Redis held the payloads of events still waiting to be processed (`e:<event_id>:<project_id>` keys, 54 GB).

What caused it, briefly:
- A flood of one repeated error filled the task queue.
- Redis's default save rules triggered near-constant 54 GB snapshots, which saturated the disk and slowed the taskbroker.
- Relay project configs expired after 2h20m because the rebuild jobs were stuck in the queue, so Relay stopped forwarding events.

What fixed it:
1. Add an inbound filter for the noisy error.
2. `redis-cli CONFIG SET save "3600 1"` to only make snapshots every hour.
3. Rebuild Relay configs directly in `sentry shell` by calling `invalidate_project_config(public_key=...)` for each ProjectKey.
4. Raise `SENTRY_TASKWORKER_CONCURRENCY` from 4 to 12.

I had previouly tried `flushall` here but ended up in the same situation a few days later.

It might be worth changing the default Redis config to include `save 3600 1` to avoid the snapshots using too much disk IO?
